neo/Consensus/ConsensusService.cs

```diff
@@ -289,17 +348,141 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
+                for (int i = 0, j = 0; i < context.Validators.Length && j < context.M; i++)
+                    if (context.FinalSignatures[i] != null)
+                    {
+                        sc.AddSignature(contract, context.Validators[i], context.FinalSignatures[i]);
+                        j++;
+                    }
+                sc.Verifiable.Witnesses = sc.GetWitnesses();
+                block.Transactions = context.TransactionHashes.Select(p => context.Transactions[p]).ToArray();
+                Log($"relay block: height={context.BlockIndex} hash={block.Hash}");
+
+                system.LocalNode.Tell(new LocalNode.Relay { Inventory = block });
+            }
+        }
+
+        private void OnRenegeration(ConsensusPayload payload, Renegeration message)
+        {
+            Log($"{nameof(OnRenegeration)}: height={payload.BlockIndex} hash={context.MakeHeader().Hash.ToString()} view={message.ViewNumber} numberOfPartialSignatures={message.SignedPayloads.Count(p => p != null)} index={payload.ValidatorIndex}");
+
+            //
+            /// <summary>
+            /// Time for checking if speaker really signed this payload
+            /// </summary>
+
+            /*
+            // TODO - Remove signature from payload and check Primary
+            public byte[] PrimaryPrepReqSignature = message.SignedPayloads[context.PrimaryIndex];
+            message.PrepReqSignature = new byte[64];
+            payload.Data = message.ToArray();
+            if (!Crypto.Default.VerifySignature(payload.GetHashData(), context.SignedPayloads[payload.ValidatorIndex], context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+            {
+                context.SignedPayloads[payload.ValidatorIndex] = null;
+                return;
+            }
+            /// <summary>
+            /// These next 2 lines could be removed, because payload is not anymore used
+            /// it was already saved before changed in the context.PreparePayload... However, let keep things clean for now
+            /// </summary>
+            message.PrepReqSignature = context.SignedPayloads[payload.ValidatorIndex];
+            payload.Data = message.ToArray(); 
+
+            */
+
+
+            /// <summary>
+            /// Time for checking all Backups
+            /// </summary>
+            uint nValidSignatures = 0;
+            for (int i = 0; i < message.SignedPayloads.Length; i++)
+                if (message.SignedPayloads[i] != null && i != message.PrepareRequestPayload.ValidatorIndex)
+                    if (!Crypto.Default.VerifySignature(message.PrepareRequestPayload.GetHashData(), message.SignedPayloads[i], context.Validators[i].EncodePoint(false)))
+                    {
+                        Log($"Regerating {i} paylod:{message.PrepareRequestPayload.ValidatorIndex} lenght:{message.SignedPayloads.Length} is being set to null");
+                        message.SignedPayloads[i] = null;
+                    }
             {
```

Comment:
Is there a "**else**" missing here ? or more braces

Reply:
Surely missing!
Thanks, @wanglongfei88.

This Regeneration part was not yet tested. We need a precise environment and adapt our script to generate a fork in this new strategy.